// edgedb/dist/syntax/select.mjs
import _std from "@generated/modules/std";
import { Cardinality, cardinalityUtil, ExpressionKind, TypeKind, } from "../reflection";
import { $expressionify } from "./path";
import { $queryify } from "./query";
export const ASC = "ASC";
export const DESC = "DESC";
export const EMPTY_FIRST = "EMPTY FIRST";
export const EMPTY_LAST = "EMPTY LAST";
export function is(expr, shape) {
    const mappedShape = {};
    for (const [key, value] of Object.entries(shape)) {
        mappedShape[key] = {
            __kind__: ExpressionKind.PolyShapeElement,
            __polyType__: expr,
            __shapeElement__: value,
        };
    }
    return mappedShape;
}
function computeFilterCardinality(expr, cardinality, base) {
    var _a, _b, _c;
    let card = cardinality;
    const filter = expr;
    const baseIsObjectExpr = ((_a = base === null || base === void 0 ? void 0 : base.__element__) === null || _a === void 0 ? void 0 : _a.__kind__) === TypeKind.object;
    const filterExprIsEq = filter.__kind__ === ExpressionKind.Operator &&
        filter.__name__ === "std::=";
    const arg0 = (_b = filter === null || filter === void 0 ? void 0 : filter.__args__) === null || _b === void 0 ? void 0 : _b[0];
    const arg1 = (_c = filter === null || filter === void 0 ? void 0 : filter.__args__) === null || _c === void 0 ? void 0 : _c[1];
    const argsExist = !!arg0 && !!arg1 && !!arg1.__cardinality__;
    const arg0IsUnique = (arg0 === null || arg0 === void 0 ? void 0 : arg0.__exclusive__) === true;
    const newCard = arg1.__cardinality__ === Cardinality.One ||
        arg1.__cardinality__ === Cardinality.AtMostOne
        ? Cardinality.AtMostOne
        : arg1.__cardinality__ === Cardinality.Empty
            ? Cardinality.Empty
            : cardinality;
    if (baseIsObjectExpr && filterExprIsEq && argsExist && arg0IsUnique) {
        if (arg0.__kind__ === ExpressionKind.PathLeaf) {
            const arg0ParentMatchesBase = arg0.__parent__.type.__element__.__name__ ===
                base.__element__.__name__;
            if (arg0ParentMatchesBase) {
                card = newCard;
            }
        }
        else if (arg0.__kind__ === ExpressionKind.PathNode) {
            const parent = arg0.__parent__;
            if (parent === null) {
                const arg0MatchesBase = arg0.__element__.__name__ === base.__element__.__name__;
                if (arg0MatchesBase) {
                    card = newCard;
                }
            }
            else {
                const arg0ParentMatchesBase = (parent === null || parent === void 0 ? void 0 : parent.type.__element__.__name__) === base.__element__.__name__;
                if (arg0ParentMatchesBase) {
                    card = newCard;
                }
            }
        }
    }
    return card;
}
function handleModifiers(modifiers, rootExpr) {
    const mods = { ...modifiers };
    let card = rootExpr.__cardinality__;
    if (mods.filter) {
        card = computeFilterCardinality(mods.filter, card, rootExpr);
    }
    if (mods.order) {
        const orderExprs = Array.isArray(mods.order) ? mods.order : [mods.order];
        mods.order = orderExprs.map(expr => typeof expr.__element__ === "undefined"
            ? expr
            : { expression: expr });
    }
    if (mods.offset) {
        mods.offset =
            typeof mods.offset === "number" ? _std.int64(mods.offset) : mods.offset;
    }
    if (mods.limit) {
        let expr = mods.limit;
        if (typeof expr === "number") {
            expr = _std.int64(expr);
        }
        else if (expr.__kind__ === ExpressionKind.Set) {
            expr = expr.__exprs__[0];
        }
        mods.limit = expr;
        if (expr.__kind__ === ExpressionKind.Literal) {
            const literalExpr = expr;
            if (literalExpr.__value__ === 1) {
                card = Cardinality.AtMostOne;
            }
            else if (literalExpr.__value__ === 0) {
                card = Cardinality.Empty;
            }
        }
    }
    return { modifiers: mods, cardinality: card };
}
function updateFunc(shape) {
    return $expressionify($queryify({
        __kind__: ExpressionKind.Update,
        __element__: this.__element__,
        __cardinality__: this.__cardinality__,
        __expr__: this,
        __shape__: shape,
    }));
}
function deleteFunc() {
    return $expressionify($queryify({
        __kind__: ExpressionKind.Delete,
        __element__: this.__element__,
        __cardinality__: this.__cardinality__,
        __expr__: this,
    }));
}
export function $selectify(expr) {
    Object.assign(expr, {
        update: updateFunc.bind(expr),
        delete: deleteFunc.bind(expr),
    });
    return $queryify(expr);
}
export function select(...args) {
    const [expr, shapeGetter] = typeof args[0].__element__ !== "undefined"
        ? args
        : [_std.FreeObject, () => args[0]];
    if (!shapeGetter) {
        if (expr.__element__.__kind__ === TypeKind.object) {
            const objectExpr = expr;
            return $expressionify($selectify({
                __kind__: ExpressionKind.Select,
                __element__: {
                    __kind__: TypeKind.object,
                    __name__: `${objectExpr.__element__.__name__}`,
                    __pointers__: objectExpr.__element__.__pointers__,
                    __shape__: { id: true },
                },
                __cardinality__: objectExpr.__cardinality__,
                __expr__: objectExpr,
                __modifiers__: {},
            }));
        }
        else {
            return $expressionify($queryify({
                __kind__: ExpressionKind.Select,
                __element__: expr.__element__,
                __cardinality__: expr.__cardinality__,
                __expr__: expr,
                __modifiers__: {},
            }));
        }
    }
    const objExpr = expr;
    const { modifiers: mods, shape, scope, expr: selectExpr, } = resolveShape(shapeGetter, objExpr);
    if (selectExpr) {
        return $expressionify($queryify({
            __kind__: ExpressionKind.Select,
            __element__: selectExpr.__element__,
            __cardinality__: cardinalityUtil.multiplyCardinalities(selectExpr.__cardinality__, objExpr.__cardinality__),
            __expr__: selectExpr,
            __modifiers__: {},
            __scope__: scope,
        }));
    }
    const { modifiers, cardinality } = handleModifiers(mods, objExpr);
    return $expressionify($selectify({
        __kind__: ExpressionKind.Select,
        __element__: {
            __kind__: TypeKind.object,
            __name__: `${objExpr.__element__.__name__}`,
            __pointers__: objExpr.__element__.__pointers__,
            __shape__: shape,
        },
        __cardinality__: cardinality,
        __expr__: expr,
        __modifiers__: modifiers,
        __scope__: expr !== _std.FreeObject ? scope : undefined,
    }));
}
function resolveShape(shapeGetter, expr) {
    const modifiers = {};
    const shape = {};
    const scope = $expressionify({
        ...expr,
        __cardinality__: Cardinality.One,
    });
    const selectShape = typeof shapeGetter === "function" ? shapeGetter(scope) : shapeGetter;
    if (typeof selectShape.__kind__ !== "undefined") {
        return { expr: selectShape, shape, modifiers, scope };
    }
    for (const [key, value] of Object.entries(selectShape)) {
        if (key === "filter" ||
            key === "order" ||
            key === "offset" ||
            key === "limit") {
            modifiers[key] = value;
        }
        else {
            shape[key] = resolveShapeElement(key, value, scope);
        }
    }
    return { shape, modifiers, scope };
}
function resolveShapeElement(key, value, scope) {
    var _a, _b;
    if ((typeof value === "function" &&
        ((_a = scope.__element__.__pointers__[key]) === null || _a === void 0 ? void 0 : _a.__kind__) === "link") ||
        (typeof value === "object" &&
            typeof value.__kind__ === "undefined")) {
        const childExpr = scope[key];
        const { shape: childShape, scope: childScope } = resolveShape(value, childExpr);
        return {
            __kind__: ExpressionKind.Select,
            __element__: {
                __kind__: TypeKind.object,
                __name__: `${childExpr.__name__}`,
                __pointers__: childExpr.__pointers__,
                __shape__: childShape,
            },
            __cardinality__: scope.__element__.__pointers__[key].cardinality,
            __expr__: childExpr,
            __modifiers__: {},
            __scope__: childScope,
        };
    }
    else if (((_b = value) === null || _b === void 0 ? void 0 : _b.__kind__) === ExpressionKind.PolyShapeElement) {
        const polyElement = value;
        const polyScope = scope.$is(polyElement.__polyType__);
        return {
            __kind__: ExpressionKind.PolyShapeElement,
            __polyType__: polyScope,
            __shapeElement__: resolveShapeElement(key, polyElement.__shapeElement__, polyScope),
        };
    }
    else {
        return value;
    }
}
