// edgedb/dist/ifaces.js
"use strict";
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2020-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderCodes = exports.ALLOW_MODIFICATIONS = exports.OPTIONS = exports.INNER = exports.BorrowReason = void 0;
var BorrowReason;
(function (BorrowReason) {
    BorrowReason["TRANSACTION"] = "transaction";
    BorrowReason["QUERY"] = "query";
    BorrowReason["CLOSE"] = "close";
})(BorrowReason = exports.BorrowReason || (exports.BorrowReason = {}));
exports.INNER = Symbol("INNER");
exports.OPTIONS = Symbol("OPTIONS");
exports.ALLOW_MODIFICATIONS = Symbol("ALLOW_MODIFICATIONS");
exports.HeaderCodes = {
    implicitLimit: 0xff01,
    implicitTypenames: 0xff02,
    implicitTypeids: 0xff03,
    allowCapabilities: 0xff04,
};
