// edgedb/dist/reflection/generators/generateSetImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSetImpl = void 0;
const builders_1 = require("../builders");
const functionUtils_1 = require("../util/functionUtils");
const generateObjectTypes_1 = require("./generateObjectTypes");
const generateSetImpl = ({ dir, types, casts }) => {
    const code = dir.getPath("syntax/setImpl");
    const implicitCastableRootTypes = (0, functionUtils_1.getImplicitCastableRootTypes)(casts);
    code.addImport({
        TypeKind: true,
        ExpressionKind: true,
        Cardinality: true,
        cardinalityUtil: true,
        $mergeObjectTypes: true,
    }, "edgedb/dist/reflection/index", true);
    code.addStarImport("castMaps", "../castMaps", true, ["ts", "js"]);
    code.addImport({ $expressionify: true }, "./path", true, ["ts", "js"]);
    code.writeln([
        (0, builders_1.t) `import type {
  ArrayType,
  TypeSet,
  BaseType,
  ObjectTypeSet,
  PrimitiveTypeSet,
  AnyTupleType,
} from "edgedb/dist/reflection";
import type {
  $expr_Set,
  mergeObjectTypesVariadic,
  getTypesFromExprs,
  getTypesFromObjectExprs,
  getCardsFromExprs,
  getSharedParentPrimitiveVariadic,
  getPrimitiveBaseType,
  LooseTypeSet,
} from "./set";`,
    ]);
    code.writeln([
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `type getSetTypeFromExprs<
  Exprs extends [TypeSet, ...TypeSet[]]
> = LooseTypeSet<
  getSharedParentPrimitiveVariadic<getTypesFromExprs<Exprs>>,
  cardinalityUtil.mergeCardinalitiesVariadic<getCardsFromExprs<Exprs>>
>;`,
    ]);
    code.writeln([
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `function set<Type extends BaseType>(
  type: Type
): $expr_Set<TypeSet<Type, Cardinality.Empty>>;
`,
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `function set<
  Expr extends TypeSet
>(expr: Expr): $expr_Set<Expr>;`,
    ]);
    for (const implicitRootTypeId of implicitCastableRootTypes) {
        code.writeln([
            (0, builders_1.dts) `declare `,
            (0, builders_1.t) `function set<
  Expr extends TypeSet<${(0, generateObjectTypes_1.getStringRepresentation)(types.get(implicitRootTypeId), {
                types,
                casts: casts.implicitCastFromMap,
            }).staticType}>,
  Exprs extends [Expr, ...Expr[]]
>(...exprs: Exprs): $expr_Set<getSetTypeFromExprs<Exprs>>;`,
        ]);
        code.writeln([
            (0, builders_1.dts) `declare `,
            (0, builders_1.t) `function set<
  Expr extends TypeSet<ArrayType<${(0, generateObjectTypes_1.getStringRepresentation)(types.get(implicitRootTypeId), {
                types,
                casts: casts.implicitCastFromMap,
            }).staticType}>>,
  Exprs extends [Expr, ...Expr[]]
>(...exprs: Exprs): $expr_Set<getSetTypeFromExprs<Exprs>>;`,
        ]);
    }
    code.writeln([
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `function set<
  Expr extends ObjectTypeSet,
  Exprs extends [Expr, ...Expr[]]
>(
  ...exprs: Exprs
): $expr_Set<
  LooseTypeSet<
    mergeObjectTypesVariadic<getTypesFromObjectExprs<Exprs>>,
    cardinalityUtil.mergeCardinalitiesVariadic<getCardsFromExprs<Exprs>>
  >
>;
`,
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `function set<
  Expr extends TypeSet<AnyTupleType>,
  Exprs extends [Expr, ...Expr[]]
>(...exprs: Exprs): $expr_Set<getSetTypeFromExprs<Exprs>>;
`,
        (0, builders_1.dts) `declare `,
        (0, builders_1.t) `function set<
  Expr extends PrimitiveTypeSet,
  Exprs extends
    TypeSet<getPrimitiveBaseType<Expr["__element__"]>>[]
>(
  expr: Expr,
  ...exprs: Exprs
): $expr_Set<
  TypeSet<
    getPrimitiveBaseType<Expr["__element__"]>,
    cardinalityUtil.mergeCardinalitiesVariadic<
      getCardsFromExprs<[Expr, ...Exprs]>
    >
  >
>;`,
    ]);
    code.writeln([
        (0, builders_1.r) `function set(..._exprs`,
        (0, builders_1.ts) `: any[]`,
        (0, builders_1.r) `) {
  // if arg
  //   return empty set
  // if object set
  //   merged objects
  // if primitive
  //   return shared parent of scalars
  if (
    _exprs.length === 1 &&
    Object.values(TypeKind).includes(_exprs[0].__kind__)
  ) {
    const element`,
        (0, builders_1.ts) `: BaseType`,
        (0, builders_1.r) ` = _exprs[0]`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `;
    return $expressionify({
      __kind__: ExpressionKind.Set,
      __element__: element,
      __cardinality__: Cardinality.Empty,
      __exprs__: [],
    })`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `;
  }
  const exprs`,
        (0, builders_1.ts) `: TypeSet[]`,
        (0, builders_1.r) ` = _exprs;
  if (exprs.every((expr) => expr.__element__.__kind__ === TypeKind.object)) {
    // merge object types;
    return $expressionify({
      __kind__: ExpressionKind.Set,
      __element__: exprs
        .map((expr) => expr.__element__`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `)
        .reduce($mergeObjectTypes),
      __cardinality__: cardinalityUtil.mergeCardinalitiesVariadic(
        exprs.map((expr) => expr.__cardinality__)`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `
      ),
      __exprs__: exprs,
    })`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `;
  }
  if (exprs.every((expr) => expr.__element__.__kind__ !== TypeKind.object)) {
    return $expressionify({
      __kind__: ExpressionKind.Set,
      __element__: exprs
        .map((expr) => expr.__element__`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `)
        .reduce(castMaps.getSharedParentScalar),
      __cardinality__: cardinalityUtil.mergeCardinalitiesVariadic(
        exprs.map((expr) => expr.__cardinality__)`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `
      ),
      __exprs__: exprs,
    })`,
        (0, builders_1.ts) ` as any`,
        (0, builders_1.r) `;
  }
  throw new Error(
    \`Invalid arguments to set constructor: \${(_exprs`,
        (0, builders_1.ts) ` as TypeSet[]`,
        (0, builders_1.r) `)
      .map((expr) => expr.__element__.__name__)
      .join(", ")}\`
  );
}`,
    ]);
    code.addExport("set");
};
exports.generateSetImpl = generateSetImpl;
