// edgedb/dist/reflection/generators/generateFunctionTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateFuncopTypes = exports.generateFunctionTypes = void 0;
const genutil_1 = require("../util/genutil");
const builders_1 = require("../builders");
const generateObjectTypes_1 = require("./generateObjectTypes");
const functionUtils_1 = require("../util/functionUtils");
const generateFunctionTypes = ({ dir, functions, types, casts, }) => {
    generateFuncopTypes(dir, types, casts, functions, "Function", "FuncExpr", true, (code, funcDef, args, namedArgs, returnType) => {
        code.writeln([(0, builders_1.t) `${(0, genutil_1.quote)(funcDef.name)},`]);
        code.writeln([(0, builders_1.t) `${args}`]);
        code.writeln([(0, builders_1.t) `${namedArgs}`]);
        code.writeln([(0, builders_1.t) `${returnType}`]);
    }, (code, funcName) => {
        code.writeln([(0, builders_1.r) `__name__: ${(0, genutil_1.quote)(funcName)},`]);
        code.writeln([(0, builders_1.r) `__args__: positionalArgs,`]);
        code.writeln([(0, builders_1.r) `__namedargs__: namedArgs,`]);
    });
};
exports.generateFunctionTypes = generateFunctionTypes;
function generateFuncopTypes(dir, types, casts, funcops, funcopExprKind, typeDefSuffix, optionalUndefined, typeDefGen, implReturnGen) {
    const typeSpecificities = (0, functionUtils_1.getTypesSpecificity)(types, casts);
    const implicitCastableRootTypes = (0, functionUtils_1.getImplicitCastableRootTypes)(casts);
    for (const [funcName, _funcDefs] of funcops.entries()) {
        const funcDefs = (0, functionUtils_1.expandFuncopAnytypeOverloads)((0, functionUtils_1.sortFuncopOverloads)(_funcDefs, typeSpecificities), types, casts, implicitCastableRootTypes);
        const { mod, name } = (0, genutil_1.splitName)(funcName);
        const code = dir.getModule(mod);
        code.registerRef(funcName, funcDefs[0].id);
        code.addRefsDefaultExport((0, genutil_1.getRef)(funcName, { prefix: "" }), name);
        const overloadsBuf = new builders_1.CodeBuffer();
        let overloadDefIndex = 1;
        for (const funcDef of funcDefs) {
            const { params } = funcDef;
            const hasParams = params.positional.length + params.named.length > 0;
            const namedParamsOverloads = !hasParams ||
                params.positional.length === 0 ||
                params.named.some(param => !(param.typemod === "OptionalType" || param.hasDefault))
                ? [true]
                : params.named.length > 0
                    ? [true, false]
                    : [false];
            for (const hasNamedParams of namedParamsOverloads) {
                if (funcDef.description) {
                    overloadsBuf.writeln([
                        (0, builders_1.t) `/**
 * ${funcDef.description.replace(/\*\//g, "")}
 */`,
                    ]);
                }
                const functionTypeName = (0, genutil_1.frag) `${(0, genutil_1.getRef)(funcName, {
                    prefix: "",
                })}λ${typeDefSuffix}${overloadDefIndex++ > 1 ? String(overloadDefIndex - 1) : ""}`;
                const functionTypeSig = (0, genutil_1.frag) `${functionTypeName}${hasParams
                    ? `<${[
                        ...(hasNamedParams ? ["NamedArgs"] : []),
                        ...params.positional.map(param => param.typeName),
                    ].join(", ")}>`
                    : ""};`;
                code.writeln([
                    (0, builders_1.dts) `declare `,
                    (0, builders_1.t) `type ${functionTypeName}${hasParams ? `<` : ` = _.syntax.$expr_${funcopExprKind}<`}`,
                ]);
                overloadsBuf.writeln([
                    (0, builders_1.dts) `declare `,
                    (0, builders_1.t) `function ${(0, genutil_1.getRef)(funcName, { prefix: "" })}${hasParams ? "<" : (0, genutil_1.frag) `(): ${functionTypeSig}`}`,
                ]);
                const anytypes = funcDef.anytypes;
                const anytypeParams = [];
                function getParamAnytype(paramTypeName, paramType, optional) {
                    if (!anytypes)
                        return undefined;
                    if (anytypes.kind === "castable") {
                        if (paramType.name.includes("anytype")) {
                            const path = (0, functionUtils_1.findPathOfAnytype)(paramType.id, types);
                            anytypeParams.push(optional
                                ? `${paramTypeName} extends $.TypeSet ? ${paramTypeName}${path} : undefined`
                                : `${paramTypeName}${path}`);
                        }
                        return anytypes.type;
                    }
                    else {
                        return anytypes.refName === paramTypeName
                            ? anytypes.type
                            : `_.syntax.getPrimitiveBaseType<${anytypes.refName}${anytypes.refPath}>`;
                    }
                }
                if (hasParams) {
                    code.indented(() => {
                        overloadsBuf.indented(() => {
                            if (hasNamedParams) {
                                code.writeln([(0, builders_1.t) `NamedArgs extends {`]);
                                overloadsBuf.writeln([(0, builders_1.t) `NamedArgs extends {`]);
                                code.indented(() => {
                                    overloadsBuf.indented(() => {
                                        for (const param of params.named) {
                                            const anytype = getParamAnytype(param.typeName, param.type, param.typemod === "OptionalType" || !!param.hasDefault);
                                            const paramType = (0, generateObjectTypes_1.getStringRepresentation)(param.type, {
                                                types,
                                                anytype,
                                                casts: casts.implicitCastFromMap,
                                            });
                                            const line = (0, builders_1.t) `${(0, genutil_1.quote)(param.name)}${param.typemod === "OptionalType" || param.hasDefault
                                                ? "?"
                                                : ""}: $.TypeSet<${paramType.staticType}>,`;
                                            code.writeln([line]);
                                            overloadsBuf.writeln([line]);
                                        }
                                    });
                                });
                                code.writeln([(0, builders_1.t) `},`]);
                                overloadsBuf.writeln([(0, builders_1.t) `},`]);
                            }
                            for (const param of params.positional) {
                                const anytype = getParamAnytype(param.typeName, param.type, optionalUndefined &&
                                    (param.typemod === "OptionalType" || !!param.hasDefault));
                                const paramTypeStr = (0, generateObjectTypes_1.getStringRepresentation)(param.type, {
                                    types,
                                    anytype,
                                    casts: casts.implicitCastFromMap,
                                });
                                const type = (0, genutil_1.frag) `$.TypeSet<${paramTypeStr.staticType}>`;
                                const line = (0, builders_1.t) `${param.typeName} extends ${param.kind === "VariadicParam"
                                    ? (0, genutil_1.frag) `[${type}, ...${type}[]]`
                                    : type}${optionalUndefined &&
                                    (param.typemod === "OptionalType" || param.hasDefault)
                                    ? " | undefined"
                                    : ""},`;
                                code.writeln([line]);
                                overloadsBuf.writeln([line]);
                            }
                        });
                    });
                    code.writeln([(0, builders_1.t) `> = _.syntax.$expr_${funcopExprKind}<`]);
                    overloadsBuf.writeln([(0, builders_1.t) `>(`]);
                    overloadsBuf.indented(() => {
                        if (hasNamedParams) {
                            overloadsBuf.writeln([(0, builders_1.t) `namedArgs: NamedArgs,`]);
                        }
                        for (const param of params.positional) {
                            overloadsBuf.writeln([
                                (0, builders_1.t) `${param.kind === "VariadicParam" ? "..." : ""}${param.internalName}${optionalUndefined &&
                                    (param.typemod === "OptionalType" || param.hasDefault)
                                    ? "?"
                                    : ""}: ${param.typeName}${param.kind === "VariadicParam" ? "" : ","}`,
                            ]);
                        }
                    });
                    overloadsBuf.writeln([(0, builders_1.t) `): ${functionTypeSig}`]);
                }
                code.indented(() => {
                    const returnAnytype = anytypes
                        ? anytypes.kind === "castable"
                            ? anytypeParams.length <= 1
                                ? anytypeParams[0]
                                : anytypeParams.slice(1).reduce((parent, type) => {
                                    return `${anytypes.returnAnytypeWrapper}<${parent}, ${type}>`;
                                }, anytypeParams[0])
                            : `_.syntax.getPrimitiveBaseType<${anytypes.refName}${anytypes.refPath}>`
                        : undefined;
                    const returnType = (0, generateObjectTypes_1.getStringRepresentation)(types.get(funcDef.return_type.id), {
                        types,
                        anytype: returnAnytype,
                    });
                    typeDefGen(code, funcDef, (0, genutil_1.frag) `[${params.positional
                        .map(param => `${param.kind === "VariadicParam" ? "..." : ""}${param.typeName}`)
                        .join(", ")}],`, (0, genutil_1.frag) `${hasParams && hasNamedParams ? "NamedArgs" : "{}"},`, (0, genutil_1.frag) `$.TypeSet<${returnType.staticType}, ${generateReturnCardinality(params, funcDef.return_typemod, hasNamedParams, funcDef.preserves_optionality)}>`);
                });
                code.writeln([(0, builders_1.t) `>;`]);
            }
        }
        code.writeBuf(overloadsBuf);
        code.writeln([
            (0, builders_1.r) `function ${(0, genutil_1.getRef)(funcName, { prefix: "" })}(...args`,
            (0, builders_1.ts) `: any[]`,
            (0, builders_1.r) `) {`,
        ]);
        code.indented(() => {
            code.writeln([
                (0, builders_1.r) `const {${funcDefs[0].kind ? "kind, " : ""}returnType, cardinality, args: positionalArgs, namedArgs} = _.syntax.$resolveOverload(args, _.spec, [`,
            ]);
            code.indented(() => {
                let overloadIndex = 0;
                for (const funcDef of funcDefs) {
                    if (funcDef.overloadIndex !== overloadIndex) {
                        continue;
                    }
                    overloadIndex++;
                    const { params } = funcDef;
                    function getArgSpec(param) {
                        return `{typeId: ${(0, genutil_1.quote)(param.type.id)}, optional: ${(param.typemod === "OptionalType" || !!param.hasDefault).toString()}, setoftype: ${(param.typemod === "SetOfType").toString()}, variadic: ${(param.kind === "VariadicParam").toString()}}`;
                    }
                    const argsDef = params.positional.map(param => {
                        return getArgSpec(param);
                    });
                    const namedArgsDef = params.named.length
                        ? `namedArgs: {${params.named
                            .map(param => {
                            return `${(0, genutil_1.quote)(param.name)}: ${getArgSpec(param)}`;
                        })
                            .join(", ")}}, `
                        : "";
                    code.writeln([
                        (0, builders_1.r) `{${funcDef.kind ? `kind: ${(0, genutil_1.quote)(funcDef.kind)}, ` : ""}args: [${argsDef.join(", ")}], ${namedArgsDef}returnTypeId: ${(0, genutil_1.quote)(funcDef.return_type.id)}${funcDef.return_typemod === "SingletonType"
                            ? ""
                            : `, returnTypemod: ${(0, genutil_1.quote)(funcDef.return_typemod)}`}${funcDef.preserves_optionality
                            ? `, preservesOptionality: true`
                            : ""}},`,
                    ]);
                }
            });
            code.writeln([(0, builders_1.r) `]);`]);
            code.writeln([(0, builders_1.r) `return _.syntax.$expressionify({`]);
            code.indented(() => {
                code.writeln([(0, builders_1.r) `__kind__: $.ExpressionKind.${funcopExprKind},`]);
                code.writeln([(0, builders_1.r) `__element__: returnType,`]);
                code.writeln([(0, builders_1.r) `__cardinality__: cardinality,`]);
                implReturnGen(code, funcName, funcDefs);
            });
            code.writeln([(0, builders_1.r) `})`, (0, builders_1.ts) ` as any`, (0, builders_1.r) `;`]);
        });
        code.writeln([(0, builders_1.r) `};`]);
        code.nl();
    }
}
exports.generateFuncopTypes = generateFuncopTypes;
function generateReturnCardinality(params, returnTypemod, hasNamedParams, preservesOptionality = false) {
    if (returnTypemod === "SetOfType") {
        return `$.Cardinality.Many`;
    }
    const paramCardinalities = [
        ...params.positional.map(p => ({
            ...p,
            genTypeName: p.typeName,
        })),
        ...(hasNamedParams
            ? params.named.map(p => ({
                ...p,
                genTypeName: `NamedArgs[${(0, genutil_1.quote)(p.name)}]`,
            }))
            : []),
    ].map(param => {
        if (param.typemod === "SetOfType") {
            if (preservesOptionality) {
                return `$.cardinalityUtil.overrideUpperBound<${param.genTypeName}["__cardinality__"], "One">`;
            }
            else {
                return `$.Cardinality.One`;
            }
        }
        if (param.typemod === "OptionalType" || param.hasDefault) {
            const _alias = `$.cardinalityUtil.optionalParamCardinality`;
            return `${_alias}<${param.genTypeName}>`;
        }
        if (param.kind === "VariadicParam") {
            return `$.cardinalityUtil.multiplyCardinalitiesVariadic<$.cardinalityUtil.paramArrayCardinality<${param.genTypeName}>>`;
        }
        return `${param.genTypeName}["__cardinality__"]`;
    });
    const cardinality = paramCardinalities.length
        ? paramCardinalities.length > 1
            ? paramCardinalities
                .slice(1)
                .reduce((cards, card) => `$.cardinalityUtil.multiplyCardinalities<${cards}, ${card}>`, paramCardinalities[0])
            : paramCardinalities[0]
        : "$.Cardinality.One";
    return returnTypemod === "OptionalType" && !preservesOptionality
        ? `$.cardinalityUtil.overrideLowerBound<${cardinality}, 'Zero'>`
        : cardinality;
}
